stdlib/regex.gr

```diff
@@ -0,0 +1,2983 @@
+        [hd] => hd,
+        mList => RESequence(mList, List.some(needsBacktrack, mList))
+      }
+    }
+  }
+}
+
+let makeREAlts = (rx1, rx2, limitC) => {
+  match((rx1, rx2)) {
+    ((RENever, _)) => rx2,
+    ((_, RENever)) => rx1,
+    ((RERange(r1), RERange(r2))) => makeRERange(rangeUnion(r1, r2), limitC),
+    ((RERange(r1), RELiteral(c2))) => makeRERange(rangeAdd(r1, Char.code(c2)), limitC),
+    ((RELiteral(c1), RERange(r2))) => makeRERange(rangeAdd(r2, Char.code(c1)), limitC),
+    ((RELiteral(c1), RELiteral(c2))) => makeRERange(rangeAdd(rangeAdd([], Char.code(c1)), Char.code(c2)), limitC),
+    _ => REAlts(rx1, rx2)
+  }
+}
+
+let makeRECut = (rx, nStart, numN) => {
+  RECut(rx, nStart, numN, needsBacktrack(rx))
+}
+
+let makeREConditional = (tst, pces1, pces2, nStart, numN) => {
+  let nb = needsBacktrack(pces1) || match(pces2) {
+    None => false,
+    Some(p2) => needsBacktrack(p2)
+  }
+  REConditional(tst, pces1, pces2, nStart, numN, nb)
+}
+
+/*
+
+=========================
+REGEX PARSING DEFINITIONS
+=========================
+
+*/
+
+// Range parsing ("[a-z]")
+
+// [TODO] (#769) When byte-based regexes are supported, we'll need another limit of 255 for those.
+let rangeLimit = 0x10FFFF
+
+// These are snake-cased to avoid confusion with their capitalized counterparts
+
+let range_d = () => {
+  rangeAddSpan([], Char.code('0'), Char.code('9'))
+}
+
+let range_w = () => {
+  rangeAdd(rangeAddSpan(rangeAddSpan(range_d(), Char.code('a'), Char.code('z')), Char.code('A'), Char.code('Z')), Char.code('_'))
+}
+
+let range_s = () => {
+  // newline, tab, page, return
+  rangeAdd(rangeAdd(rangeAdd(rangeAdd(rangeAdd([], Char.code(' ')), 9), 10), 12), 13)
+}
+
+let rec parseRangeNot = (buf: RegExBuf) => {
+  if (!more(buf)) {
+    Err(parseErr(buf, "Missing closing `]`", 0))
+  } else {
+    match(peek(buf)) {
+      Err(e) => Err(e),
+      Ok('^') => {
+        ignore(eat(buf, '^'))
+        match(parseRange(buf)) {
+          Err(e) => Err(e),
+          Ok(rng) => Ok(rangeInvert(rng, rangeLimit))
+        }
+      },
+      Ok(_) => parseRange(buf)
+    }
+  }
+},
+
+parseRange = (buf: RegExBuf) => {
+  if (!more(buf)) {
+    Err(parseErr(buf, "Missing closing `]`", 0))
+  } else {
+    match(peek(buf)) {
+      Err(e) => Err(e),
+      Ok(']') => {
+        ignore(eat(buf, ']'))
+        match(parseRangeRest(buf, [], None, None)) {
+          Err(e) => Err(e),
+          Ok(rng) => Ok(rangeAdd(rng, Char.code(']')))
+        }
+      },
+      Ok('-') => {
+        ignore(eat(buf, '-'))
+        match(parseRangeRest(buf, [], None, None)) {
+          Err(e) => Err(e),
+          Ok(rng) => Ok(rangeAdd(rng, Char.code('-')))
+        }
+      },
+      Ok(_) => parseRangeRest(buf, [], None, None)
+    }
+  }
+},
+
+parseClass = (buf: RegExBuf) => {
+  if (!more(buf)) {
+    Err("no chars") // caught in handler (we use a Result to cleanly mesh with the Result type below)
+  } else {
+    match(peek(buf)) {
+      Err(e) => Err(e),
```

Comment:
You can see below (in `parseRangeRest`) that we actually ignore what's in the error altogether and just replace it with a generic "invalid class" error. More importantly, this `Err` from `peek` should never happen given that `more(buf)` must be true inside of this `if` branch. This could probably be streamlined into just the `match(peek(buf))` (without the enclosing `if`), but it looked just like this as a result of the port from Racket.

Reply:
I'd love to see the `match(peek(buf))` changes land in this PR, but I also understand if you want to put the cleanup off until later since it is a port.